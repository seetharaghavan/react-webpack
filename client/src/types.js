export const  REQUEST_LOG_IN = "REQUEST_LOG_IN"; 
export const  LOG_IN = "LOG_IN"; 
export const  LOG_OUT = "LOG_OUT"; 
export const  LOG_IN_FAILED = "LOG_IN_FAILED"; 

export const  ADD_NOTE = "ADD_NOTE";
export const  EDIT_NOTE = "EDIT_NOTE"; 
export const  REMOVE_NOTE = "REMOVE_NOTE";

export const  GET_NOTES = "GET_NOTES"; 
export const  SELECT_NOTE = 'SELECT_NOTE'; 
export const  ERROR_NOTE = 'ERROR_NOTE'; 

export const  NOTE_REQUEST = "NOTE_REQUEST"; 

export const  SIDEBAR_TOGGLE = 'SIDEBAR_TOGGLE'; 

export const  NOTE_FORM_RESET = "NOTE_FORM_RESET"; 

export const  BASE_URL = "http://localhost:4000"; 