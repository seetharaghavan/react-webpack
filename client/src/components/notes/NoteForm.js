import React from 'react'; 
import { useDispatch, useSelector } from 'react-redux';
import { addNewNote, updateNote } from '../../services/actions';
import { NOTE_FORM_RESET, SIDEBAR_TOGGLE } from '../../types';

const NoteForm = ({noteState}) => {
  const [noteObj, setNoteObj] = React.useState({ title: "", body: "" });

  const [errors, setErrors] = React.useState({ field: null, message: "" });

  const dispatch = useDispatch(); 

  const sidebarState = useSelector(state => state.sidebarState); 


  React.useEffect(() => {
    if(noteState && noteState.selectedNote){
        const {title, body} = noteState.selectedNote; 
        setNoteObj(prev => ({title, body})) 
    } else {
      setNoteObj(prev => ({title: '', body: ''}))
    }
  }, [noteState])

  const onFormSubmit = (e) => {
    e.preventDefault();
    if(ValidateForm(noteObj)){
        let notes = {
            ...noteObj
        }
        if(noteState && noteState.selectedNote) {
            notes.id = noteState.selectedNote.id; 
            dispatch(updateNote(notes)); 
        } else{
            notes.id = Math.floor(Math.random() * 100000); 
            dispatch(addNewNote(notes)); 
        }

        dispatch({type: SIDEBAR_TOGGLE, payload: true}); 
        dispatch({type: NOTE_FORM_RESET}); 
        
        setNoteObj(prev => ({
            ...prev,
            title: '',
            body: ''
        }));
    } else {
        return; 
    }
  };

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setNoteObj((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const ValidateForm = (noteObj) => {
    let isValid = false;
    if (noteObj.title.length === 0) {
      setErrors((prev) => ({
        ...prev,
        field: "title",
        message: "Add title",
      }));
      return isValid;
    } else if (noteObj.body.length === 0) {
      setErrors((prev) => ({
        ...prev,
        field: "body",
        message: "Add a brief description",
      }));
      return isValid;
    } else {
      isValid = true;
      setErrors((prev) => ({
        ...prev,
        field: null,
        message: "",
      }));
    }
    return isValid;
  };

  return (
    <div className={`note-area collapse ${!sidebarState.sidebarOpen ? 'show' : ''}`}>
      <form onSubmit={onFormSubmit}>
        <div className="form-group text-right">
          <button className="btn btn-light" onClick={(e) => {
            e.preventDefault(); 
            dispatch({type: NOTE_FORM_RESET}); 
            setNoteObj({title: '', body: ''}); 
          }}>
            Add New
          </button>
        </div>
        <div className="form-group">
          <label>Title</label>
          <input
            type="text"
            className="form-control"
            placeholder="Enter title"
            value={noteObj.title}
            name="title"
            onChange={handleInputChange}
            data-testid="note-title"
            required
          />
          {
              errors.field === 'title' && <small className="text-danger">{errors.message}</small>
          }
        </div>
        <div className="form-group">
          <label>Body</label>
          <textarea
            className="form-control"
            rows={15}
            name="body"
            value={noteObj.body}
            placeholder="Enter description"
            onChange={handleInputChange}
            data-testid="note-body"
            required
          />
           {
              errors.field === 'body' && <small className="text-danger">{errors.message}</small>
          }
        </div>
        <div className="form-group text-right">
          <button type="submit" data-testid="note-submit" className="btn btn-primary">
            Save
          </button>
        </div>
      </form>
    </div>
  );
};

export default NoteForm; 
