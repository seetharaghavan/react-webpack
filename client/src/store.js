import {createStore, combineReducers, applyMiddleware} from 'redux'; 
import ReduxThunk  from 'redux-thunk'; 
import authReducer from './reducers/auth.reducer';
import noteReducer from './reducers/note.reducer';
import sideBarReducer from './reducers/sidebar.reducer';

const store = createStore( 
    combineReducers({
        authState: authReducer,
        noteState: noteReducer,
        sidebarState: sideBarReducer
    }),
    {},
    applyMiddleware(ReduxThunk)
)

export default store; 

