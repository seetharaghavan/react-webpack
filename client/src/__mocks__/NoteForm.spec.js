import React from "react";
import renderer from "react-test-renderer";
import localStorage from "./localstorage";
import { Provider } from "react-redux";
import store from "../store";
import NoteForm from "../components/notes/NoteForm";
import { render, fireEvent, waitFor, cleanup } from "@testing-library/react";
import { ADD_NOTE } from "../types";

const component = (
  <Provider store={store}>
    <NoteForm />
  </Provider>
);

afterEach(cleanup);

beforeAll(() => {
	global.console.error = jest.fn((...args) => {
		if (typeof args[0] === 'string' && args[0].includes('It looks like you\'re using the wrong act() around your test interactions')) {
	      return
	    }
	    return originalError.call(console, args)
	});
});

afterAll(() => {
  	global.console.error.mockRestore();
});

describe("Basic Ui", () => {
  //basic ui
  it("should render", async () => {
    const tree = await renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have  2 input components", () => {
    const { getByTestId } = render(component);
    expect(getByTestId("note-submit")).toHaveTextContent("Save");
    expect(getByTestId("note-title")).toBeRequired();
    expect(getByTestId("note-body")).toBeRequired();
  });
});

describe("Notes Form behaviour", () => {
    let testnote = {
        title: 'Test title',
        body: 'Test body'
    }
  it("empty fields", async () => {
    const { getByTestId, getByText } = render(component);

    fireEvent.change(getByTestId("note-title"), {
      target: { value: "" },
    });

    fireEvent.change(getByTestId("note-body"), {
      target: { value: "" },
    });
    fireEvent.click(getByTestId("note-submit"));
    expect(getByText("Add title")).toBeInTheDocument();
  });

  it("empty body", async () => {
    const { getByTestId, getByText } = render(component);

    fireEvent.change(getByTestId("note-title"), {
      target: { value: testnote.title },
    });

    fireEvent.change(getByTestId("note-body"), {
      target: { value: "" },
    });

    fireEvent.click(getByTestId("note-submit"));

    expect(getByText("Add a brief description")).toBeInTheDocument();
  });

  it("ideal add scenario", async () => {
    const { getByTestId, getByText, queryByText } = render(component);
    fireEvent.change(getByTestId("note-title"), {
      target: { value: testnote.title },
    });

    fireEvent.change(getByTestId("note-body"), {
      target: { value: testnote.body },
    });

    fireEvent.click(getByTestId("note-submit"));
    store.dispatch({type: ADD_NOTE, payload: {
        id: Math.floor(Math.random() * 10000),
        title: testnote.title,
        body: testnote.body
    }})
    await waitFor(() => {
        expect(getByTestId('note-title')).toHaveValue(''); 
        expect(getByTestId('note-body')).toHaveValue(''); 
        expect(queryByText("Add title")).not.toBeInTheDocument();
        expect(queryByText("Add a brief description")).not.toBeInTheDocument();
    });
  });
});
