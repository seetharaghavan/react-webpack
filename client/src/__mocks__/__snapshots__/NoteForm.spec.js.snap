// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Ui should render 1`] = `
<div
  className="note-area collapse "
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group text-right"
    >
      <button
        className="btn btn-light"
        onClick={[Function]}
      >
        Add New
      </button>
    </div>
    <div
      className="form-group"
    >
      <label>
        Title
      </label>
      <input
        className="form-control"
        data-testid="note-title"
        name="title"
        onChange={[Function]}
        placeholder="Enter title"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Body
      </label>
      <textarea
        className="form-control"
        data-testid="note-body"
        name="body"
        onChange={[Function]}
        placeholder="Enter description"
        required={true}
        rows={15}
        value=""
      />
    </div>
    <div
      className="form-group text-right"
    >
      <button
        className="btn btn-primary"
        data-testid="note-submit"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
